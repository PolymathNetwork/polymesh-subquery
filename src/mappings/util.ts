import { Codec } from "@polkadot/types/types";
import { u8aToString } from "@polkadot/util";
/**
 * @returns a javascript object built using an `iterable` of keys and values.
 * Values are mapped by the map parameter
 */
export const fromEntries = <K extends string | number, V, V2>(
  iterable: Iterable<[K, V]>,
  map: (v: V, i: number, k: K) => V2
): Partial<Record<K, V2>> => {
  const res: Partial<Record<K, V2>> = {};

  let i = 0;
  for (const [k, v] of iterable) {
    res[k] = map(v, i, k);
    i++;
  }
  return res;
};

export const camelToSnakeCase = (str: string): string =>
  str.replace(/[A-Z]/g, (letter) => `_${letter.toLowerCase()}`);

export const capitalizeFirstLetter = (str: string): string =>
  str[0].toUpperCase() + str.slice(1);

export const removeNullChars = (s: string): string => s.replace(/\0/g, "");

/**
 * @returns the index of the first top level comma in `text` which is a string with nested () and <>.
 * This is meant for rust types, for example "Map<Map<(u8,u32),String>,bool>"" would return 24.
 * @param text
 */
export const findTopLevelCommas = (
  text: string,
  exitOnFirst = false
): number[] => {
  let nestedLevel = 0;
  const commas = [];
  let i = 0;

  for (const char of text) {
    switch (char) {
      case "(":
      case "{":
      case "<": {
        nestedLevel++;
        break;
      }
      case ")":
      case "}":
      case ">": {
        nestedLevel--;
        break;
      }
      case ",": {
        if (nestedLevel === 1) {
          if (exitOnFirst) {
            return [i];
          } else {
            commas.push(i);
          }
        }
      }
    }
    i++;
  }
  if (commas.length === 0) {
    throw new Error(
      `No top level comma found in ${text}, it probably isn't a map`
    );
  }
  return commas;
};

export const getOrDefault = <K, V>(
  map: Map<K, V>,
  key: K,
  getDefault: () => V
): V => {
  const v = map.get(key);
  if (v !== undefined) {
    return v;
  } else {
    const def = getDefault();
    map.set(key, def);
    return def;
  }
};

export const serializeTicker = (item: Codec): string => {
  return removeNullChars(u8aToString(item.toU8a()));
};
