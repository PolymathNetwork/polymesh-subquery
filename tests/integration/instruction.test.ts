import { gql } from "@apollo/client/core";
import { getApolloClient } from "../util";
const { query } = getApolloClient();

describe("getInstructions", () => {
  test("test getting instructions", async () => {
    const q = {
      query: gql`
        query {
          instructions(first: 10) {
            totalCount
            nodes {
              blockId
              status
              venueId
              settlementType
              addresses
            }
          }
        }
      `,
    };

    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();
    expect(subquery?.data).toMatchSnapshot();
  });
});
