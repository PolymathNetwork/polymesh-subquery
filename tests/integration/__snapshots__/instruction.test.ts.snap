// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getInstructions test getting instructions 1`] = `
Object {
  "instructions": Object {
    "__typename": "InstructionsConnection",
    "nodes": Array [
      Object {
        "__typename": "Instruction",
        "addresses": Array [
          "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "5DcLNmYXCC8J1hLHJsbtPFNX7kSDx4YTbgGkNYcHUKdvbgU9",
        ],
        "blockId": 259,
        "settlementType": "SettleOnAffirmation",
        "status": "Executed",
        "venueId": "1",
      },
      Object {
        "__typename": "Instruction",
        "addresses": Array [
          "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "5FA8HQPMNmidk6X2h9dfZUEivwRt6ju4cgtpHPEcbB62veMW",
          "5EXPTATZUFcAR9ao5qp9UVaYPoprDpQZFhuVpyibFhA3gmCY",
          "5Hmn2MX8PEd7rBfmXywhAfMvr77FkVy94PyMGncWabDbFVJ7",
          "5DUtBYagbSiwaVdgK54ka26dtdykVn5M71DduDdw64M659EW",
        ],
        "blockId": 288,
        "settlementType": "SettleOnAffirmation",
        "status": "Executed",
        "venueId": "2",
      },
      Object {
        "__typename": "Instruction",
        "addresses": Array [
          "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "5FcXLkN8fkcr65qg8ZJpjZ4Aeqk8HBvXKdDmSK2zUwB4EJsc",
        ],
        "blockId": 324,
        "settlementType": "SettleOnAffirmation",
        "status": "Executed",
        "venueId": "3",
      },
      Object {
        "__typename": "Instruction",
        "addresses": Array [
          "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "5Ev8nm5CBFm3FRvjso95EbCmVM44WRcZRyH4jgXv6CpLvxTa",
        ],
        "blockId": 352,
        "settlementType": "SettleOnAffirmation",
        "status": "Executed",
        "venueId": "5",
      },
      Object {
        "__typename": "Instruction",
        "addresses": Array [
          "5CaWrWYw133scCHRLGBkQv9ssdFFnanPSZhQHmo8ARMeK79Z",
          "5GqANhefumSBPko8oLdKAH8No6Wk7iN4kzgesvqxyQMupfSu",
        ],
        "blockId": 377,
        "settlementType": "SettleOnAffirmation",
        "status": "Executed",
        "venueId": "6",
      },
      Object {
        "__typename": "Instruction",
        "addresses": Array [
          "5CaWrWYw133scCHRLGBkQv9ssdFFnanPSZhQHmo8ARMeK79Z",
        ],
        "blockId": 383,
        "settlementType": "SettleOnAffirmation",
        "status": "Executed",
        "venueId": "7",
      },
      Object {
        "__typename": "Instruction",
        "addresses": Array [
          "5GqANhefumSBPko8oLdKAH8No6Wk7iN4kzgesvqxyQMupfSu",
        ],
        "blockId": 384,
        "settlementType": "SettleOnAffirmation",
        "status": "Executed",
        "venueId": "7",
      },
    ],
    "totalCount": 7,
  },
}
`;
